### PR TITLE
Implement endpoint for getting dependent objects alerts in AlertContr…

### LINKED ISSUE #186: Implement dependent objects' alerts in the Alert API

### PR DESCRIPTION
**Pull request by MarselAhmetov**
…oller

Resolve #186

### CODE REVIEW

**Review by Leshe4ka — APPROVED** [review]

**Comment by DementevNikita** [comment]
@MarselAhmetov it'd be great if you could send the SQL that jOOQ generates

File: odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/repository/AlertRepositoryImpl.java [hunk 1]
```diff
@@ -119,6 +125,68 @@
...
+            .build();
+    }
+
+    private List<String> getObjectsOddrnsByOwner(final long ownerId) {
+        return dslContext.select(DATA_ENTITY.ODDRN)
+            .from(DATA_ENTITY)
+            .leftJoin(OWNERSHIP).on(DATA_ENTITY.ID.eq(OWNERSHIP.DATA_ENTITY_ID))
+            .where(OWNERSHIP.OWNER_ID.eq(ownerId))
+            .fetchStreamInto(String.class)
+            .collect(Collectors.toList());
```
Comment (damirabdul) [inline odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/repository/AlertRepositoryImpl.java:169]: you can use toList() from java 16, which a bit better

File: odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/repository/AlertRepositoryImpl.java [hunk 2]
```diff
@@ -119,6 +125,68 @@
...
+            .from(LINEAGE)
+            .where(LINEAGE.CHILD_ODDRN.in(ownOddrns))
+            .unionAll(dslContext
+                .select(LINEAGE.asterisk())
+                .from(LINEAGE)
+                .join(cteName).on(LINEAGE.CHILD_ODDRN.eq(field("t.parent_oddrn", String.class)))));
+
+        return name("t2")
+            .as(dslContext.withRecursive(cte)
+                .selectDistinct(field(cte.getName() + ".child_oddrn", String.class))
```
Comment (damirabdul) [inline odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/repository/AlertRepositoryImpl.java:186]: it's better to use field(name("table_name", "column_name")). Thus you can avoid String column names

File: odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/repository/AlertRepositoryImpl.java [hunk 3]
```diff
@@ -138,6 +206,21 @@ public long countByOwner(final long ownerId) {
...
+    @Override
+    public long countDependentObjectsAlerts(final long ownerId) {
+        final List<String> ownOddrns = getObjectsOddrnsByOwner(ownerId);
+
+        final CommonTableExpression<Record1<String>> cte = getChildOddrnsLinageByOwnOddrnsCte(ownOddrns);
+
+        return dslContext.with(cte)
+            .select(countDistinct(ALERT.ID))
+            .from(ALERT)
+            .join(cte.getName()).on(field(cte.getName() + ".child_oddrn", String.class).eq(ALERT.DATA_ENTITY_ODDRN))
```
Comment (damirabdul) [inline odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/repository/AlertRepositoryImpl.java:218]: the same as above

File: odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/repository/AlertRepositoryImpl.java [hunk 4]
```diff
@@ -119,6 +125,68 @@
...
+            .join(DATA_ENTITY_SUBTYPE).on(DATA_ENTITY_SUBTYPE.ID.eq(DATA_ENTITY.SUBTYPE_ID))
+            .where(DATA_ENTITY.ODDRN.notIn(ownOddrns))
+            .groupBy(selectFields)
+            .fetchStream()
+            .map(this::mapRecord)
+            .collect(Collectors.toList());
+
+        return Page.<AlertDto>builder()
+            .data(data)
+            .hasNext(true)
```
Comment (damirabdul) [inline odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/repository/AlertRepositoryImpl.java:158]: Why do we always set hasNext as true and total as 0? @DementevNikita i see, that it copy-pasted from listByOwner method, is it expected?

File: odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/repository/AlertRepositoryImpl.java [hunk 5]
```diff
@@ -119,6 +125,68 @@
             .collect(Collectors.toList());
     }
 
+    @Override
+    public Page<AlertDto> listDependentObjectsAlerts(final int page, final int size, final long ownerId) {
```
Comment (DementevNikita) [inline odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/repository/AlertRepositoryImpl.java:129]: It seems you have recursive CTE inside another CTE. Let's have a chained CTEs and not nested

**Review by DementevNikita — CHANGES_REQUESTED** [review]

File: odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/repository/AlertRepositoryImpl.java (hunk 2 above)
Comment (DementevNikita) [reply odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/repository/AlertRepositoryImpl.java:186]: Let's not use strings at all, we have jOOQ constants. This way we can be sure the app won't compile if we change lineage fields

File: odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/repository/AlertRepositoryImpl.java (hunk 3 above)
Comment (DementevNikita) [reply odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/repository/AlertRepositoryImpl.java:218]: https://github.com/opendatadiscovery/odd-platform/pull/281#discussion_r761895681

**Review by DementevNikita — APPROVED** [review]

**Pull request merged**